File: go/internal/scoring/scoring.go
// Package scoring computes field-level accuracy of a predicted Invoice against
// ground truth (the benchmark harness core).
package scoring

import (
	"strings"
	"time"

	"github.com/shopspring/decimal"

	"invoice-extractor/internal/schema"
)

// ScalarFields is the fixed display/scoring order for per-field accuracy.
var ScalarFields = []string{
	"vendor_name", "transaction_datetime", "currency", "category", "total_amount",
}

type InvoiceScore struct {
	FieldsTotal       int
	FieldsCorrect     int
	Accuracy          float64
	PerField          map[string]bool
	LineItemPrecision float64
	LineItemRecall    float64
	LineItemF1        float64
}

func Score(pred, gold schema.Invoice) InvoiceScore {
	pf := map[string]bool{
		"vendor_name":          eqStr(pred.VendorName, gold.VendorName),
		"transaction_datetime": eqDate(pred.TransactionDatetime, gold.TransactionDatetime),
		"currency":             eqStr(pred.Currency, gold.Currency),
		"category":             eqStr(pred.Category, gold.Category),
		"total_amount":         eqDec(pred.TotalAmount, gold.TotalAmount),
	}
	correct := 0
	for _, v := range pf {
		if v {
			correct++
		}
	}
	total := len(pf)
	p, r, f1 := scoreLineItems(pred.LineItems, gold.LineItems)
	return InvoiceScore{
		FieldsTotal:       total,
		FieldsCorrect:     correct,
		Accuracy:          float64(correct) / float64(total),
		PerField:          pf,
		LineItemPrecision: p,
		LineItemRecall:    r,
		LineItemF1:        f1,
	}
}

func norm(s string) string { return strings.ToLower(strings.Join(strings.Fields(s), " ")) }

func eqStr(a, b *string) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return norm(*a) == norm(*b)
}

func eqDate(a, b *time.Time) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return a.Format("2006-01-02") == b.Format("2006-01-02")
}

func eqDec(a, b *decimal.Decimal) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return a.Equal(*b)
}

func lineEq(p, g schema.LineItem) bool {
	if norm(p.Description) != norm(g.Description) {
		return false
	}
	return eqDec(p.Amount, g.Amount)
}

func scoreLineItems(pred, gold []schema.LineItem) (precision, recall, f1 float64) {
	if len(pred) == 0 && len(gold) == 0 {
		return 1, 1, 1
	}
	used := make([]bool, len(pred))
	matched := 0
	for _, g := range gold {
		for i, p := range pred {
			if used[i] {
				continue
			}
			if lineEq(p, g) {
				used[i] = true
				matched++
				break
			}
		}
	}
	if len(pred) > 0 {
		precision = float64(matched) / float64(len(pred))
	}
	if len(gold) > 0 {
		recall = float64(matched) / float64(len(gold))
	}
	if precision+recall > 0 {
		f1 = 2 * precision * recall / (precision + recall)
	}
	return precision, recall, f1
}

File: python/app/scoring.py
"""Field-level accuracy scoring: compare a predicted Invoice to ground truth.

Used by scripts/bench.py (the decision tool) and tests/test_scoring.py. Scalar
fields are scored exact-match (strings normalized for whitespace/case); line items
get precision/recall via greedy description+amount matching.
"""

from __future__ import annotations

from pydantic import BaseModel

from app.schema import Invoice, LineItem

# Scalar fields scored for exact match (line_items scored separately). The expense
# tracker keeps only what/where/when/how-much + category; total_amount is the computed
# sum of line item prices.
SCALAR_FIELDS = (
    "vendor_name",
    "transaction_datetime",
    "currency",
    "category",
    "total_amount",
)


class InvoiceScore(BaseModel):
    fields_total: int
    fields_correct: int
    accuracy: float
    per_field: dict[str, bool]
    line_item_precision: float
    line_item_recall: float
    line_item_f1: float


def score_invoice(pred: Invoice, gold: Invoice) -> InvoiceScore:
    per_field = {f: _scalar_eq(getattr(pred, f), getattr(gold, f)) for f in SCALAR_FIELDS}
    total = len(SCALAR_FIELDS)
    correct = sum(per_field.values())
    p, r, f1 = _score_line_items(pred.line_items, gold.line_items)
    return InvoiceScore(
        fields_total=total,
        fields_correct=correct,
        accuracy=correct / total if total else 1.0,
        per_field=per_field,
        line_item_precision=p,
        line_item_recall=r,
        line_item_f1=f1,
    )


def _norm(s: str) -> str:
    return " ".join(s.split()).casefold()


def _scalar_eq(a, b) -> bool:
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    if isinstance(a, str) and isinstance(b, str):
        return _norm(a) == _norm(b)
    return a == b


def _line_eq(p: LineItem, g: LineItem) -> bool:
    return _norm(p.description or "") == _norm(g.description or "") and p.amount == g.amount


def _score_line_items(
    pred: list[LineItem], gold: list[LineItem]
) -> tuple[float, float, float]:
    if not pred and not gold:
        return 1.0, 1.0, 1.0
    used: set[int] = set()
    matched = 0
    for g in gold:
        for i, p in enumerate(pred):
            if i in used:
                continue
            if _line_eq(p, g):
                used.add(i)
                matched += 1
                break
    precision = matched / len(pred) if pred else 0.0
    recall = matched / len(gold) if gold else 0.0
    f1 = (
        2 * precision * recall / (precision + recall)
        if (precision + recall) > 0
        else 0.0
    )
    return precision, recall, f1
